xarray/core/computation.py

```diff
@@ -396,8 +399,37 @@ def _unpack_dict_tuples(
     return out
 
 
+def _check_core_dims(signature, variable_args, name):
+    """
+    Chcek if an arg has all the core dims required by the signature.
+
+    Slightly awkward design, of returning the error message. But we want to
+    give a detailed error message, which requires inspecting the variable in
+    the inner loop.
+    """
+    missing = []
+    for core_dims, variable_arg in zip(signature.input_core_dims, variable_args):
+        # Check whether all the dims are on the variable. Note that we need the
+        # `hasattr` to check for a dims property, to protect against the case where
+        # a numpy array is passed in.
+        if hasattr(variable_arg, "dims") and set(core_dims) - set(variable_arg.dims):
+            missing += [[variable_arg, core_dims]]
+    if missing:
+        message = f"Missing core dims from `{name}` on variable(s):\n\n"
+        for variable_arg, core_dims in missing:
+            message += f"Missing {set(core_dims) - set(variable_arg.dims)} on\n{variable_arg}\n\n"
```

Comment:
This may be clearer if we could add an "argument number".

So "Missing ['x'] in argument number 2"

Reply:
OK I rewrote it from scratch after some sleep, WDYT? Please continue refining, your suggestions have been excellent

(These error messages take time but I actually think it's really nice when they're nice — as someone returning to xarray as being occasionally befuddled. A good exercise in usability!)

Reply:
Here's the full error message:

```
E                   ValueError: Missing core dims {'y'} from arg number 2 on a variable named `x_z`:
E                   <xarray.Variable (x: 2, z: 3)>
E                   array([[0, 1, 2],
E                          [3, 4, 5]])
E
E                   Either add the core dimension, or if passing a dataset alternatively pass `on_missing_core_dim` as `copy` or `drop`.
```

(Would love a pytest-accept-like solution to snapshot these, much easier to both iterate on and understand as a reviewer than a partial regex)

Reply:
Perhaps a slight reordering? This is quite minor so feel free to ignore

"Missing core dims {'y'} on a variable named `x_z` in arg number 2"